### PR TITLE
#161382366 Add Search Functionality and Custom filtering based on parameters

### PR DESCRIPTION
**Pull request by johnkegz**
#### What does this PR do?
Compares develop with ft-search-functionality-161382366
#### Description of Task to be completed?
Users would be able to search for articles and would be able to filter articles by author, article title and tag/keywords, this would ensure a user only gets an article that matches ONLY the criteria that matches the criteria for the article they want.
#### How should this be manually tested?
Run the Heroku link in the browser, click on search, select the search type, you will be prompted with a an input box top input your details, press submit.
#### Any background context you want to provide?
Users should be able to search for articles.
#### What are the relevant pivotal tracker stories?
#161382366
#### Screenshots (if appropriate)
screen shot 2018-12-14 at 00 42 50
screen shot 2018-12-14 at 00 43 04
screen shot 2018-12-14 at 00 43 21
screen shot 2018-12-14 at 00 43 40
screen shot 2018-12-14 at 00 45 00

### CODE REVIEW

File: src/containers/search/SearchForm.jsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    this.setState({ [e.target.name]: e.target.value });
+  };
+
+  onSubmit = e => {
+    const type = this.state.type;
+    const searchData = this.state.searchData;
+    this.props.history.push(`/searchresults/${type}/${searchData}`);
+  };
+
+  myFunction = e => {
```
Comment (shakirandagire) [inline src/containers/search/SearchForm.jsx]: @johnkegz Could you use a more meaningful name for this function

File: src/containers/search/SearchForm.jsx [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    e.preventDefault();
+    this.setState({ [e.target.name]: e.target.value });
+  };
+
+  helperFunction1() {
```
Comment (shakirandagire) [inline src/containers/search/SearchForm.jsx]: Could you also get a more meaningful name for this function

File: src/containers/search/SearchResults.jsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import PropTypes from "prop-types";
+
+export class SearchResults extends Component {
+  componentDidMount() {
+    this.props.getSearchData(
+      this.props.match.params.type,
+      this.props.match.params.searchData
+    );
+  }
+  helperFUnction(){
```
Comment (shakirandagire) [inline src/containers/search/SearchResults.jsx]: Could write a meaningful name for this function

**Review by emukungu — APPROVED** [review]
The implementation of nested routes looks good.

**Review by cdvx — APPROVED** [review]
Looks okay to me, other than some commented code, everything else is OK👍🏾

**Review by cdvx — COMMENTED** [review]
Looks good, though there are many changed files. Good things is most are snapshots👍🏾

**Review by bozicschucky — APPROVED** [review]
Looks good but after receiving feedback you should avoid changing too many files.

File: src/containers/search/SearchForm.jsx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+              <option>search</option>
+              <option value="title">Article title</option>
+              <option value="tags">Tag</option>
+              <option value="author_name">Authors name</option>
+            </select>
+          </div>
+          <input
+            type="text"
+            placeholder="Search"
+            style={this.state.type? { display: "block", color: "white" }: { display: "none", color: "white" }}
```
Comment (shakirandagire) [inline src/containers/search/SearchForm.jsx]: Could you remove inline css

File: src/containers/search/SearchResults.jsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+};
+
+export class SearchResults extends Component {
+  componentDidMount() {
+    this.props.getSearchData(
+      this.props.match.params.type,
+      this.props.match.params.searchData
+    );
+  }
+  articleDiv = (id,name,title) => {
```
Comment (shakirandagire) [inline src/containers/search/SearchResults.jsx]: Could you make this function more descriptive

File: src/containers/search/SearchResults.jsx [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+              <div className="col-lg-12 col-md-12 mb-12">
+                <div className="card">{this.articleDiv(data.id,data.author.username, data.title)}</div>
+              </div>
+            </section>
+          </div>
+        </main>
+      </div>
+    ));
+  }
+  notFound() {
```
Comment (shakirandagire) [inline src/containers/search/SearchResults.jsx]: Could you add what is not found

**Comment by bozicschucky** [comment]
@johnkegz  resolve the conflicts for this PR.

File: src/containers/search/SearchForm.jsx (hunk 1 above)
Comment (johnkegz) [reply src/containers/search/SearchForm.jsx]: @shakirandagire I have worked on this.

File: src/containers/search/SearchForm.jsx (hunk 2 above)
Comment (johnkegz) [reply src/containers/search/SearchForm.jsx]: @shakirandagire I have worked on this.

File: src/containers/search/SearchResults.jsx (hunk 3 above)
Comment (johnkegz) [reply src/containers/search/SearchResults.jsx]: @shakirandagire I have worked on this.

File: src/containers/search/SearchForm.jsx (hunk 4 above)
Comment (johnkegz) [reply src/containers/search/SearchForm.jsx]: @shakirandagire I have worked on this.

File: src/containers/search/SearchResults.jsx (hunk 5 above)
Comment (johnkegz) [reply src/containers/search/SearchResults.jsx]: @shakirandagire I have worked on this.

File: src/containers/search/SearchResults.jsx (hunk 6 above)
Comment (johnkegz) [reply src/containers/search/SearchResults.jsx]: @shakirandagire I have worked on this.

**Comment by emukungu** [comment]
Looks ok

**Pull request merged**